### PR TITLE
ENG-178

### PR DESCRIPTION
**Pull request by Connorbelez**
### TL;DR

Added comprehensive chaos tests and financial invariant stress tests for the cash ledger system, plus regression verification to ensure the ownership ledger remains untouched.

### What changed?

Added three new test files to verify cash ledger resilience and system integrity:

1. **`chaosTests.test.ts`** - Tests 5 chaos scenarios from the tech design:
   - Cash receipts arriving before obligation settlement
   - Duplicate webhook idempotency handling
   - Settlement callbacks after cancellation
   - Concurrent settlement creating overpayments
   - Dispersal mutation failures with reconciliation detection

2. **`financialInvariantStress.test.ts`** - Stress tests for financial invariants:
   - Conservation through reversal and re-collection cycles
   - CONTROL:ALLOCATION account balancing with partial reversals
   - Negative balance enforcement boundaries
   - Point-in-time reconstruction with 50+ entries
   - Idempotent replay verification

3. **`regressionVerification.test.ts`** - Meta-tests to enforce REQ-244:
   - Git diff verification that no ownership ledger files were modified
   - Structural checks for expected ownership ledger files

### How to test?

Run the test suite to verify all new tests pass:

```bash
bun run test
```

The new tests cover:
- Out-of-order webhook delivery scenarios
- Duplicate webhook handling via idempotency keys
- Financial invariant preservation under stress
- Regression protection for the ownership ledger

### Why make this change?

These tests fulfill ENG-178 requirements to verify the cash ledger behaves correctly under distributed system conditions like out-of-order delivery, duplicates, and failures. The chaos tests ensure the system handles real-world webhook timing issues, while stress tests verify financial invariants hold under edge cases. The regression verification enforces the hard constraint that the ownership ledger remains completely untouched, protecting existing functionality.

## Summary by CodeRabbit

* **Tests**
  * Added comprehensive test suites for payment ledger validation, including chaos scenario testing (out-of-order transactions, duplicate requests, concurrent operations), financial invariant stress testing (balance conservation, allocation correctness, negative balance controls), and regression verification to ensure system integrity is maintained under edge cases.

### CODE REVIEW

File: convex/payments/cashLedger/__tests__/chaosTests.test.ts [hunk 1]
```diff
@@ -0,0 +1,564 @@
...
+			return postCashReceiptForObligation(ctx, {
+				obligationId,
+				amount: 100_000,
+				idempotencyKey: sharedKey,
+				source: SYSTEM_SOURCE,
+			});
+		});
+		expect(secondEntry).toBeDefined();
+		expect(secondEntry).toHaveProperty("entry");
+
```
[the remark this answers is not in the document]
Comment (Connorbelez) [reply convex/payments/cashLedger/__tests__/chaosTests.test.ts]: Fixed in f7c7176 — both `firstEntry` and `secondEntry` assertions now use `.not.toBeNull()` to unambiguously verify the receipt was journaled.

File: convex/payments/cashLedger/__tests__/chaosTests.test.ts [hunk 2]
```diff
@@ -0,0 +1,564 @@
...
+			const result = await postCashReceiptForObligation(ctx, {
+				obligationId,
+				amount: 100_000,
+				idempotencyKey: buildIdempotencyKey(
+					"cash-received-chaos-1",
+					obligationId
+				),
+				source: SYSTEM_SOURCE,
+			});
+			expect(result).toBeDefined();
```
[the remark this answers is not in the document]
Comment (Connorbelez) [reply convex/payments/cashLedger/__tests__/chaosTests.test.ts]: Fixed — changed to `expect(result).not.toBeNull()` since `postCashReceiptForObligation` can return `null` and `toBeDefined()` would still pass for `null`.

File: convex/payments/cashLedger/__tests__/regressionVerification.test.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+				message.includes("unknown revision") ||
+				message.includes("bad revision")
+			) {
+				console.warn(
+					"Skipping git diff check — 'main' branch not available in this environment"
+				);
+				return;
+			}
+			// All other git failures should surface, not be silently swallowed
+			throw new Error(`Git diff failed unexpectedly: ${message}`);
```
[the remark this answers is not in the document]
Comment (Connorbelez) [reply convex/payments/cashLedger/__tests__/regressionVerification.test.ts]: **Already fixed** in commit `2e505a30` ("responding to feedback").

The current code implements all the suggested improvements:
- Detects CI via `process.env.CI` / `process.env.GITHUB_ACTIONS`
- Prefers `GITHUB_BASE_REF` in CI, falls back to `origin/main`, then `main`
- Fetches `origin/main --depth=1` as last resort in CI
- Hard-fails in CI if no base ref is available (no silent skip)
- Still skips gracefully in local dev when no base ref exists

File: convex/payments/cashLedger/__tests__/financialInvariantStress.test.ts [hunk 4]
```diff
@@ -0,0 +1,619 @@
...
+			const journalSettled = await getJournalSettledAmountForObligation(
+				ctx as unknown as QueryCtx,
+				obligationId
+			);
+			expect(
+				journalSettled,
+				"Journal settled amount should match net receipts"
+			).toBe(100_000n);
+		});
+	});
```
[the remark this answers is not in the document]
Comment (Connorbelez) [reply convex/payments/cashLedger/__tests__/financialInvariantStress.test.ts]: Renamed the test from "conservation holds after reversal and re-collection" to "conservation of BORROWER_RECEIVABLE balance through accrue→receipt→reversal→re-collection cycle" to accurately reflect that it validates BORROWER_RECEIVABLE balance consistency and journal-derived settled amounts, not full conservation of value through settlement allocation/dispersal. Fixed in 329f0b5b.

File: convex/payments/cashLedger/__tests__/financialInvariantStress.test.ts [hunk 5]
```diff
@@ -0,0 +1,619 @@
...
+			const account = await ctx.db.get(lenderPayable._id);
+			if (!account) {
+				throw new Error("LENDER_PAYABLE account not found after REVERSAL");
+			}
+			const balance = getCashAccountBalance(account as CashLedgerAccountDoc);
+			expect(
+				balance,
+				"LENDER_PAYABLE should be negative after REVERSAL"
+			).toBeLessThan(0n);
+		});
```
[the remark this answers is not in the document]
Comment (Connorbelez) [reply convex/payments/cashLedger/__tests__/financialInvariantStress.test.ts]: Added a post-reversal enforcement assertion: after the REVERSAL pushes LENDER_PAYABLE negative, the test now attempts a non-REVERSAL LENDER_PAYOUT_SENT entry (amount=1000) debiting the already-negative LENDER_PAYABLE and asserts it throws the negative balance error. This confirms that enforcement resumes for non-REVERSAL postings even when the account is in a negative state from a prior reversal. Fixed in 329f0b5b.
